New Entries: Java Comparator & Comparable (#832)
Codecademy/docs: content/java/concepts/comparable/comparable.md

---
Title: 'Comparable'
Description: 'The Comparable interface is used to define how a class is to be sorted.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Collections'
  - 'Data Structures'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

> The **Comparable** interface is used to define how a class is to be sorted. It is not to be confused with the [Comparator](https://www.codecademy.com/resources/docs/java/comparator) interface, which is implemented in a separate class. The Comparable interface is implemented in the class to be sorted.

Reviewer: ```suggestion
The **`Comparable`** interface is used to define how a [class](https://www.codecademy.com/resources/docs/java/classes) is to be sorted. It is not to be confused with the [`Comparator`](https://www.codecademy.com/resources/docs/java/comparator) interface, which is implemented in a separate class. The `Comparable` interface is implemented in the class to be sorted.
```

> The Comparable interface only has one method, `.compareTo()` that is used to compare the current instance of the class with another instance of the same class, and the sort is based on the return value of `.compareTo()`:

Reviewer: ```suggestion
The `Comparable` interface only has one method, `.compareTo()` that is used to compare the current instance of the class with another instance of the same class, and the sort is based on the return value of `.compareTo()`:
```

| Return Value | Means                           |
| :----------: | ------------------------------- |
|    `>= 1`    | this instance > passed instance |
|     `0`      | this instance = passed instance |
|   `<= -1`    | this instance < passed instance |

> This method determines how items are sorted by methods such as `Arrays.sort()` and `Collections.sort()`.

Reviewer: ```suggestion
This method determines how items are sorted by methods such as [`Arrays.sort()`](https://www.codecademy.com/resources/docs/java/arrays/sort) and `Collections.sort()`.
```

## Syntax

```pseudo
class MyClass implements Comparable<MyClass> {
  // Class body.
  ...
  @Override public int compareTo(MyClass value)
  {
      // Comparison Logic
      ...
      return result;
  }
}
```

Using the interface in `MyClass` requires it to implement `Comparable<MyClass>` and override the `.compareTo()` method. The `.compareTo()` method can then use any applicable logic to decide if `this` instance comes before, after, or in the same place as `value` in the sort order and return the appropriate `int` value.

## Example

The following example defines an `Employee` class that sorts based on `lastName` then on `firstName`.

```java
import java.io.*;
import java.util.*;

> 
> class Employee implements Comparable<Employee> {

Reviewer: ```suggestion

// Employee.java
class Employee implements Comparable<Employee> {
```

  String firstName;
  String lastName;

  public Employee(String first, String last)
  {
    this.firstName = first;
    this.lastName = last;
  }

  public String toString()
  {
    return "( " + lastName + ", " + firstName + " )";
  }

  @Override public int compareTo(Employee value)
  {
    if (this.lastName.compareTo(value.lastName) != 0) {
      return this.lastName.compareTo(value.lastName);
    } else {
      return this.firstName.compareTo(value.firstName);
    }
  }
}

> 
> public class SortExample {

Reviewer: ```suggestion

// SortExample.java
public class SortExample {
```

Reviewer: This might be helpful to anyone who tests this code, runs it, and gets errors. They'd need to separate into two appropriately-named `.java` files.

The `Comparable` interface only has one method, `.compareTo()` that is used to compare the current instance of the class with another instance of the same class, and the sort is based on the return value of `.compareTo()`:

| Return Value | Means                           |
| :----------: | ------------------------------- |
|    `>= 1`    | this instance > passed instance |
|     `0`      | this instance = passed instance |
|   `<= -1`    | this instance < passed instance |

This method determines how items are sorted by methods such as [`Arrays.sort()`](https://www.codecademy.com/resources/docs/java/arrays/sort) and `Collections.sort()`.

## Syntax

```pseudo
class MyClass implements Comparable<MyClass> {
  // Class body.
  ...
  @Override public int compareTo(MyClass value)
  {
      // Comparison Logic
      ...
      return result;
  }
}
```

Using the interface in `MyClass` requires it to implement `Comparable<MyClass>` and override the `.compareTo()` method. The `.compareTo()` method can then use any applicable logic to decide if `this` instance comes before, after, or in the same place as `value` in the sort order and return the appropriate `int` value.

## Example

> The following example defines an `Employee` class that sorts based on `lastName` then on `firstName`.

Reviewer 2: ```suggestion
The following example defines an `Employee` class that sorts based on `lastName` and then on `firstName`.
```

```java

> import java.io.*;

Reviewer 2: Since this example is a little long it might be a good idea to add comments just to explain what is going on.

---
Title: 'Comparable'
Description: 'The Comparable interface is used to define how a class is to be sorted.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Collections'
  - 'Data Structures'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

The **`Comparable`** interface is used to define how a [class](https://www.codecademy.com/resources/docs/java/classes) is to be sorted. It is not to be confused with the [`Comparator`](https://www.codecademy.com/resources/docs/java/comparator) interface, which is implemented in a separate class. The `Comparable` interface is implemented in the class to be sorted.

> The `Comparable` interface only has one method, `.compareTo()` that is used to compare the current instance of the class with another instance of the same class, and the sort is based on the return value of `.compareTo()`:
> 
> | Return Value | Means                           |
> | :----------: | ------------------------------- |
> |    `>= 1`    | this instance > passed instance |
> |     `0`      | this instance = passed instance |
> |   `<= -1`    | this instance < passed instance |
> 
> This method determines how items are sorted by methods such as [`Arrays.sort()`](https://www.codecademy.com/resources/docs/java/arrays/sort) and `Collections.sort()`.

Reviewer: ```suggestion
```
I think this should be incorporated into the "Syntax" section.

  {
      // Comparison Logic
      ...
      return result;
  }
}
```

Applying the `Comparable` interface to `MyClass` requires the `implements` keyword (e.g., `Comparable<MyClass>`). This interface has a single `.compareTo()` method that returns an `int` value based on whether the `value` of a current class instance (referenced by `this`) can be logically sorted with the value of another instance of the same class.

| Return Value | Meaning                         |
| :----------: | ------------------------------- |

> |    `>= 1`    | this instance > passed instance |
> |     `0`      | this instance = passed instance |
> |   `<= -1`    | this instance < passed instance |

Reviewer: ```suggestion
|    >= 1    | this instance > passed instance |
|     0      | this instance = passed instance |
|   <= -1    | this instance < passed instance |
```

Although this is return output, I'm not sure if they should necessarily be in back-ticks (to [a point made in a separate PR](https://github.com/Codecademy/docs/pull/840#discussion_r919231406)).

> Using the interface in `MyClass` requires it to implement `Comparable<MyClass>` and override the `.compareTo()` method. The `.compareTo()` method can then use any applicable logic to decide if `this` instance comes before, after, or in the same place as `value` in the sort order and return the appropriate `int` value.

Reviewer: ```suggestion
Applying the `Comparable` interface to `MyClass` requires the `implements` keyword (e.g., `Comparable<MyClass>`). This interface has a single `.compareTo()` method that returns an `int` value based on whether the `value` of a current class instance (referenced by `this`) can be logically sorted with the value of another instance of the same class.

| Return Value | Meaning                        |
| :----------: | ------------------------------- |
|    `>= 1`    | this instance > passed instance |
|     `0`      | this instance = passed instance |
|   `<= -1`    | this instance < passed instance |

This method determines how items are sorted by methods such as [`Arrays.sort()`](https://www.codecademy.com/resources/docs/java/arrays/sort) and `Collections.sort()`.
```
